updater/lib/dependabot/job.rb

```diff
@@ -306,64 +373,81 @@ def log_ignore_conditions_for(dependency)
     end
 
     # Remove this once the API always sends groups for grouped refresh jobs
+    sig { params(group: String).void }
     def override_group_to_refresh_due_to_old_defaults(group)
       @dependency_group_to_refresh = group
     end
 
     private
 
+    sig { params(dependency: Dependabot::Dependency).returns(T::Boolean) }
     def completely_ignored?(dependency)
       ignore_conditions_for(dependency).any?(Dependabot::Config::IgnoreCondition::ALL_VERSIONS)
     end
 
+    sig { void }
     def register_experiments
-      experiments.each do |name, value|
+      experiments.entries.each do |name, value|
         Dependabot::Experiments.register(name, value)
       end
     end
 
+    sig { params(name1: String, name2: String).returns(T::Boolean) }
     def name_match?(name1, name2)
       WildcardMatcher.match?(
-        name_normaliser.call(name1),
-        name_normaliser.call(name2)
+        T.must(name_normaliser).call(name1),
+        T.must(name_normaliser).call(name2)
       )
     end
 
+    sig do
+      params(requirements_update_strategy: T.nilable(String), lockfile_only: T::Boolean).returns(T.nilable(String))
+    end
     def build_update_strategy(requirements_update_strategy:, lockfile_only:)
       return requirements_update_strategy unless requirements_update_strategy.nil?
 
       lockfile_only ? "lockfile_only" : nil
     end
 
+    sig { params(source_details: T::Hash[String, T.untyped]).returns(Dependabot::Source) }
     def build_source(source_details)
       # Immediately normalize the source directory, ensure it starts with a "/"
-      directory = source_details["directory"]
+      directory = T.let(source_details["directory"], T.nilable(String))
       unless directory.nil?
         directory = Pathname.new(directory).cleanpath.to_s
-        source_details["directory"] = "/#{directory}" unless directory.start_with?("/")
+        directory = "/#{directory}" unless directory.start_with?("/")
       end
 
       Dependabot::Source.new(
-        **source_details.transform_keys { |k| k.tr("-", "_").to_sym }
+        provider: T.let(source_details["provider"], String),
+        repo: T.let(source_details["repo"], String),
+        directory: directory,
+        directories: T.let(source_details["directories"], T.nilable(T::Array[String])),
+        branch: T.let(source_details["branch"], T.nilable(String)),
+        commit: T.let(source_details["commit"], T.nilable(String)),
+        hostname: T.let(source_details["hostname"], T.nilable(String)),
+        api_endpoint: T.let(source_details["api-endpoint"], T.nilable(String))
       )
     end
 
     # Provides a Dependabot::Config::UpdateConfig objected hydrated with
     # relevant information obtained from the job definition.
     #
     # At present we only use this for ignore rules.
+    sig { returns(Dependabot::Config::UpdateConfig) }
     def update_config
-      return @update_config if defined? @update_config
```

Comment:
Is it intentional that we're no longer memoizing this? The operation doesn't look necessarily expensive but maybe we shouldn't lump in logical changes with the Sorbet ones, in case we hit any issues in prod it'll be easier to diagnose?

Reply:
Ah, looks like we're assigning it in the initializer and then no longer calling this method, is that something Sorbet wants us to do? We could make the change smaller by calling this `def calculate_update_config` or similar and creating an `attr_reader :update_config`, that way the calling code remains unchanged

Reply:
It does gently push you towards doing things in the initializer because using `defined?` means that it needs to be `nilable`.